external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala

```diff
@@ -18,34 +18,58 @@
 package org.apache.spark.sql.kafka010
 
 import java.{util => ju}
+import java.util.concurrent.{ConcurrentMap, TimeUnit}
 import javax.annotation.concurrent.GuardedBy
 
+import com.google.common.cache._
+import org.apache.kafka.clients.producer.KafkaProducer
+import scala.collection.JavaConverters._
 import scala.collection.immutable.SortedMap
 import scala.collection.mutable
-
-import org.apache.kafka.clients.producer.KafkaProducer
+import scala.util.control.NonFatal
 
 import org.apache.spark.internal.Logging
 
 private[kafka010] object CachedKafkaProducer extends Logging {
 
   private type Producer = KafkaProducer[Array[Byte], Array[Byte]]
 
-  @GuardedBy("this")
-  private val cacheMap = new mutable.HashMap[String, Producer]()
+  private val cacheExpireTimeout: Long = System.getProperty("spark.kafka.guava.cache.timeout",
```

Comment:
I wonder if this should be an option of the sink? I may be running multiple streams in a single cluster with different trigger intervals which would require different configs.